yearn/contracts/SaddleStrategy.sol

```diff
@@ -0,0 +1,395 @@
+    ///         available by the liquidation.
+    /// @return loss Amount of the loss.
+    function liquidatePosition(uint256 amountNeeded)
+        internal
+        override
+        returns (uint256 liquidatedAmount, uint256 loss)
+    {
+        uint256 wantBalance = balanceOfWant();
+        if (wantBalance < amountNeeded) {
+            liquidatedAmount = withdrawSome(amountNeeded.sub(wantBalance));
+            liquidatedAmount = liquidatedAmount.add(wantBalance);
+            loss = amountNeeded.sub(liquidatedAmount);
+        } else {
+            liquidatedAmount = amountNeeded;
+        }
+    }
+
+    /// @notice This method is defined in the BaseStrategy contract and is meant
+    ///         to liquidate everything and return the amount that got freed.
+    ///         This strategy implements the aforementioned behavior by
+    ///         withdrawing all vault's underlying tokens from the tBTC v2 Saddle
+    ///         reward pool.
+    /// @dev This function is used during emergency exit instead of prepareReturn
+    ///      to liquidate all of the strategy's positions back to the vault.
+    /// @return amountFreed Amount that got freed.
+    function liquidateAllPositions()
+        internal
+        override
+        returns (uint256 amountFreed)
+    {
+        ILPRewardsTBTCV2Saddle(lpRewardsTbtcV2Saddle).withdraw(balanceOfPool());
+
+        // Yearn docs doesn't specify clear enough what exactly should be
+        // returned here. It may be either the total balance after
+        // withdrawAllAndUnwrap or just the amount withdrawn. Currently opting
+        // for the former because of
+        // https://github.com/yearn/yearn-vaults/pull/311#discussion_r625588313.
+        // Also, usage of this result in the harvest method in the BaseStrategy
+        // seems to confirm that.
+        return balanceOfWant();
+    }
+
+    /// @notice This method is defined in the BaseStrategy contract and is meant
+    ///         to do anything necessary to prepare this strategy for migration,
+    ///         such as transferring any reserve or LP tokens, CDPs, or other
+    ///         tokens or stores of value. This strategy implements the
+    ///         aforementioned behavior by withdrawing all vault's underlying
+    ///         tokens from the tBTC v2 Saddle reward pool.
+    /// @param newStrategy Address of the new strategy meant to replace the
+    ///        current one.
+    function prepareMigration(address newStrategy) internal override {
+        // Withdraw all the LP tokens from the tBTC v2 Saddle reward pool.
+        // These tokens will be transferred to the new strategy in the `migrate`
+        // function.
+        ILPRewardsTBTCV2Saddle(lpRewardsTbtcV2Saddle).withdraw(balanceOfPool());
+        // Get all the earned KEEP tokens and transfer them to the new strategy.
+        ILPRewardsTBTCV2Saddle(lpRewardsTbtcV2Saddle).getReward();
```

Comment:
While writing system tests, let's make sure the reward is actually obtained. Maybe we should `getReward` before `withdraw`?

Reply:
I don't think it is a problem we call `getReward` after `withdraw`.
The `LPRewardsTBTCSaddle` contract even has an [exit](https://github.com/keep-network/keep-ecdsa/blob/96cf2541764d019237906c4e0fa7051d15f6b6cd/solidity/contracts/LPRewards.sol#L143) function that calls these two functions in the same order.
But I will check this behaviour in a system test.

Reply:
Ok good!